skbio/tree/_bme.py

```diff
@@ -0,0 +1,352 @@
+                                            \-rat
+
+    Notice that, unlike neighbor joining, the tree is rooted at a taxa/leaf node.
+    This will allow it to have nearest neighbor interchange performed on it without
+    needing to re-root the tree.
+
+    """
+    if dm.shape[0] < 3:
+        raise ValueError(
+            "Distance matrix must be at least 3x3 to "
+            "generate a minimum evolution tree."
+        )
+
+    # count the number of taxa
+    n = len(dm.ids)
+
+    # create initial nodes
+    root_name, node_a_name, node_b_name = dm.ids[0], dm.ids[1], dm.ids[2]
+    root = TreeNode(root_name)
+    root_child = TreeNode("")
+    node_a = TreeNode(node_a_name)
+    node_b = TreeNode(node_b_name)
+    root.append(root_child)
+    root_child.append(node_a)
+    root_child.append(node_b)
+
+    # loop over rest of nodes for k = 4 to n
+    for k in range(3, n):
+        taxa = root.count(tips=True) + 1
+        # calculate average distance matrix between subtrees of T_(k-1)
+        # Note that this could be replaced with an updating step to an
+        # initially computed average distance matrix, as noted in GME,
+        # however the updating step requires all subtrees to be updated
+        # in contrast to GME, so instead the matrix is recomputed after
+        # every iteration.
+        adm = _balanced_average_matrix(root, dm)
+        # create node for taxa k
+        node_k_name = dm.ids[k]
+        connecting_node = TreeNode("")
+        node_k = TreeNode(node_k_name)
+        connecting_node.append(node_k)
+        # connecting_node = TreeNode.read(["(%s);" % (node_k_name)])
+        # node_k = connecting_node.find(node_k_name)
+        # create average distance lists for subtrees of T_(k-1)
+        ordered = list(root.postorder(include_self=False))
+        lowerlist = _balanced_lower(ordered, node_k, dm)
+        upperlist = _balanced_upper(ordered, node_k, taxa, dm)
+        # Initialize the list of edge parent nodes for computation.
+        # The tuple is a pair of a TreeNode object and the value of the tree
+        # length for attaching at the edge consisting of the node and its parent.
+        # Here, we start with setting the initial length value of the edge
+        # defined by the root and its unique descendant to 0.
+        edge_list = [(root_child, 0)]
+        for a, i in edge_list:
+            if a.is_tip():
+                continue
+            a1, a2 = a.children
+            value1 = _balanced_attach_length(
+                a1, lowerlist, upperlist, ordered, taxa, adm
+            )
+            value2 = _balanced_attach_length(
+                a2, lowerlist, upperlist, ordered, taxa, adm
+            )
+            e1 = (a1, value1 + i)
+            e2 = (a2, value2 + i)
+            edge_list.append(e1)
+            edge_list.append(e2)
+        # find the edge with minimum length after edge attachment
+        minimum_child = sorted(edge_list, key=itemgetter(1))[0][0]
```

Comment:
Repeated call of `itemgetter(1)` is expensive. You may add a line `key = itemgetter(1)` outside the for loop, and use it during iteration.